src/StatusesHandler.php

```diff
@@ -0,0 +1,270 @@
+     * @param string                $screenOptionsKey Option key name to save screen statuses config.
+     * @param string                $statusesEndpoint Endpoint for saving statuses.
+     * @param EventManagerInterface $eventManager     The event manager.
+     * @param EventFactoryInterface $eventFactory     The event factory.
+     */
+    public function __construct(
+        Traversable $statuses,
+        Traversable $hiddenStatuses,
+        Traversable $statusesLabels,
+        $screenOptionsKey,
+        $statusesEndpoint,
+        $eventManager,
+        $eventFactory
+    ) {
+        $this->statuses       = $this->_normalizeArray($statuses);
+        $this->hiddenStatuses = $this->_normalizeArray($hiddenStatuses);
+        $this->statusesLabels = $this->_normalizeArray($statusesLabels);
+
+        $this->screenOptionsKey = $screenOptionsKey;
+        $this->statusesEndpoint = $statusesEndpoint;
+
+        $this->_setEventManager($eventManager);
+        $this->_setEventFactory($eventFactory);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        /* @var $event EventInterface */
+        $event = func_get_arg(0);
+
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        $event->setParams(array_merge(
+            $event->getParams(),
+            $this->_getStatusesParams()
+        ));
+    }
+
+    /**
+     * Attach events for status handling.
+     *
+     * @since [*next-version*]
+     */
+    public function run()
+    {
+        $this->_attach('eddbk_bookings_visible_statuses', function ($event) {
+            $event->setParams([
+                'statuses' => $this->_getVisibleStatuses($this->statuses, $this->hiddenStatuses),
+            ]);
+        });
+
+        /*
+         * Attach status endpoint.
+         */
+        $this->_attach('wp_ajax_set_' . $this->screenOptionsKey, function () {
+            $data = json_decode(file_get_contents('php://input'), true);
+            $statuses = $data['statuses'];
+            $this->_setScreenStatuses($this->screenOptionsKey, $statuses);
+        });
+    }
+
+    /**
+     * Get status related parameters to attach on event.
+     *
+     * @since [*next-version*]
+     *
+     * @return array
+     */
+    protected function _getStatusesParams()
+    {
+        return [
+            /*
+             * All available statuses in application.
+             */
+            'statuses' => $this->_getTranslatedStatuses($this->statuses, $this->statusesLabels),
+
+            /*
+             * Statuses that enabled for filtering bookings.
+             */
+            'screenStatuses' => $this->_getScreenStatuses($this->screenOptionsKey, $this->statuses),
```

Comment:
This is a filter, that should be made more generic. See [card](https://github.com/orgs/RebelCode/projects/2#card-9959215).